c++/src/kj/async-io.h

```diff
@@ -136,6 +136,60 @@ class AsyncOutputStream: private AsyncObject {
   // multiple times without canceling the previous promises.
 };
 
+class AsyncBufferedInputStream {
+  // Does not inherit from InputStream so this can act as a mixin to allow implementing
+  // AsyncBufferedIoStreamWrapper
+
+public:
+  virtual ~AsyncBufferedInputStream() noexcept(false);
+
+  ArrayPtr<const byte> getReadBuffer();
+  // TODO(now): No skip for AsyncInputStream -- I guess we could implement an async form?
```

Comment:
You could add `skip()` specifically to the buffered interface, and specify for now that it's an error to skip more than the current buffer. We could consider adding it to `AsyncInputStream` later but I'm not sure it's worth it.